als/src/CuMFJNIInterface.cpp

```diff
@@ -24,107 +24,108 @@
+		  printf("%d ", body[i]);
 		}
-		env->DeleteLocalRef(factorsPerBlock);
-	}
-	// get a pointer to the raw input data, pinning them in memory
-	csrRowIndexHostPtr = (jint*) env->GetPrimitiveArrayCritical(csrRow, 0);
-	csrColIndexHostPtr = (jint*) env->GetPrimitiveArrayCritical(csrCol, 0);
-	csrValHostPtr = (jfloat*) env->GetPrimitiveArrayCritical(csrVal, 0);
+		printf("\n");
+		*/
+		//calculate X from thetaT
+		float* thetaTHost;
+		cudacall(cudaMallocHost( (void** ) &thetaTHost, n * f * sizeof(thetaTHost[0])) );
+		//to be returned
+		float* XTHost;
+		cudacall(cudaMallocHost( (void** ) &XTHost, m * f * sizeof(XTHost[0])) );
+		
+		int numSrcBlocks = env->GetArrayLength(sortedSrcFactors);
+		//WARNING: ReleaseFloatArrayElements and DeleteLocalRef are important; 
+		//Otherwise result is correct but performance is bad
+		int index = 0;
+		for(int i = 0; i < numSrcBlocks; i++){
+			jobject factorsPerBlock = env->GetObjectArrayElement(sortedSrcFactors, i);
+			int numFactors = env->GetArrayLength((jobjectArray)factorsPerBlock);
+			for(int j = 0; j < numFactors; j++){
+				jobject factor = env->GetObjectArrayElement((jobjectArray)factorsPerBlock, j);
+				jfloat *factorfloat = (jfloat *) env->GetPrimitiveArrayCritical( (jfloatArray)factor, 0);	
+				memcpy(thetaTHost + index*f, factorfloat, sizeof(float)*f);
+				index ++;
+				env->ReleasePrimitiveArrayCritical((jfloatArray)factor, factorfloat, 0);
+				env->DeleteLocalRef(factor);
+			}
+			env->DeleteLocalRef(factorsPerBlock);
+		}
+		// get a pointer to the raw input data, pinning them in memory
+		csrRowIndexHostPtr = (jint*) env->GetPrimitiveArrayCritical(csrRow, 0);
+		csrColIndexHostPtr = (jint*) env->GetPrimitiveArrayCritical(csrCol, 0);
+		csrValHostPtr = (jfloat*) env->GetPrimitiveArrayCritical(csrVal, 0);
 
-	/*
-	printf("thetaTHost of len %d: \n", n*f);
-    for (int i = 0; i < n*f; i++) {
-	  printf("%f ", thetaTHost[i]);
-    }
-	printf("\n");
-	*/
-	int * d_csrRowIndex = 0;
-	int * d_csrColIndex = 0;
-	float * d_csrVal = 0;
+		/*
+		printf("thetaTHost of len %d: \n", n*f);
+		for (int i = 0; i < n*f; i++) {
+		  printf("%f ", thetaTHost[i]);
+		}
+		printf("\n");
+		*/
+		int * d_csrRowIndex = 0;
+		int * d_csrColIndex = 0;
+		float * d_csrVal = 0;
 
-	cudacall(cudaMalloc((void** ) &d_csrRowIndex,(m + 1) * sizeof(float)));
-	cudacall(cudaMalloc((void** ) &d_csrColIndex, nnz * sizeof(float)));
-	cudacall(cudaMalloc((void** ) &d_csrVal, nnz * sizeof(float)));
-	cudacall(cudaMemcpyAsync(d_csrRowIndex, csrRowIndexHostPtr,(size_t ) ((m + 1) * sizeof(float)), cudaMemcpyHostToDevice, cuda_stream));
-	cudacall(cudaMemcpyAsync(d_csrColIndex, csrColIndexHostPtr,(size_t ) (nnz * sizeof(float)), cudaMemcpyHostToDevice, cuda_stream));
-	cudacall(cudaMemcpyAsync(d_csrVal, csrValHostPtr,(size_t ) (nnz * sizeof(float)),cudaMemcpyHostToDevice, cuda_stream));
-	cudaStreamSynchronize(cuda_stream);
+		cudacall(cudaMalloc((void** ) &d_csrRowIndex,(m + 1) * sizeof(float)));

```

Comment:
Maybe in debug mode we can print every time -- it will have MANY such prints if we excessively use GPU.
In non-debug mode, may be useful we print once when degraded malloc is invoked?